compiler/src/dotty/tools/dotc/transform/LazyVals.scala

```diff
@@ -274,6 +284,274 @@ class LazyVals extends MiniPhase with IdentityDenotTransformer {
+          If(
+            ref(current).select(defn.Any_==).appliedTo(evaluating),
+            // if is Evaluating then CAS(_, Evaluating, new Waiting)
+            Block(
+              objCasFlag.appliedTo(thiz, offset, evaluating, Select(New(waiting), StdNames.nme.CONSTRUCTOR).ensureApplied) :: Nil,
+              unitLiteral
+            ),
+            // if not Evaluating
+            If(
+              ref(current).select(defn.Any_isInstanceOf).appliedToTypeTree(waiting),
+              // if is waiting
+              ref(current).select(defn.Any_asInstanceOf).appliedToTypeTree(waiting).select(lazyNme.RLazyVals.waitingAwaitRelease, _.info.paramInfoss.exists(_.size == 0)).ensureApplied,
+              If(ref(current).select(defn.Any_==).appliedTo(nullValue),
+                Return(nullLiteral, lazyInitMethodSymbol),
+                unitLiteral
+              )
+            )
+          ),
+        // if not a control state
+        Return(ref(current), lazyInitMethodSymbol)
+      )
+
+    val initBody = Block(ValDef(current, ref(target)) :: Nil, If(ref(current).equal(nullLiteral), initialize, ifNotUninitialized).withType(defn.UnitType))
+    val initMainLoop = WhileDo(EmptyTree, initBody) // becomes: while (true) do { body }
+    val initMethodDef = DefDef(lazyInitMethodSymbol, initMainLoop)
+    (accessorDef, initMethodDef)
+  }
+
+  def transformMemberDefThreadSafe(x: ValOrDefDef)(using Context): Thicket = {
+    assert(!(x.symbol is Mutable))
+    if ctx.settings.YlightweightLazyVals.value then
+      transformMemberDefThreadSafeNew(x)
+    else
+      transformMemberDefThreadSafeLegacy(x)
+  }
+
+  def transformMemberDefThreadSafeNew(x: ValOrDefDef)(using Context): Thicket = {
+    import dotty.tools.dotc.core.Types._
+    import dotty.tools.dotc.core.Flags._
+
+    val runtimeModule = "scala.runtime.LazyVals"
+    val claz = x.symbol.owner.asClass
+    val thizClass = Literal(Constant(claz.info))
+    val helperModule = requiredModule(runtimeModule)
+    var offsetSymbol: TermSymbol | Null = null
+
+    def offsetName(id: Int) = s"${StdNames.nme.LAZY_FIELD_OFFSET}${if (x.symbol.owner.is(Module)) "_m_" else ""}$id".toTermName
+    val containerName = LazyLocalName.fresh(x.name.asTermName)
+    val containerSymbol = newSymbol(claz, containerName, containerFlags, defn.ObjectType).enteredAfter(this)
+    containerSymbol.addAnnotation(Annotation(defn.VolatileAnnot)) // private @volatile var _x: AnyRef
+    containerSymbol.addAnnotations(x.symbol.annotations) // pass annotations from original definition
+    val stat = x.symbol.isStatic
+    if stat then
+      containerSymbol.setFlag(JavaStatic)
+    val getOffset =
+      if stat then
+        Select(ref(helperModule), lazyNme.RLazyVals.getStaticFieldOffset)
+      else
+        Select(ref(helperModule), lazyNme.RLazyVals.getOffsetStatic)
+    val containerTree = ValDef(containerSymbol, nullLiteral)
+
+    // create an offset for this lazy val
+    appendOffsetDefs.get(claz) match
```

Comment:
The two branches of this `match` look *very* similar. Can't they be factored out?